frostsnap_core/src/device.rs

```diff
@@ -254,129 +243,181 @@ impl FrostSigner {
                     .public_key()
                     .to_xonly_bytes();
 
-                self.state = SignerState::FrostKey {
-                    key: FrostsnapKey {
+                self.action_state = Some(SignerState::KeyGenAck {
+                    key: FrostsnapSecretKey {
                         frost_key,
                         secret_share,
                         share_index: *my_index,
-                        aux_rand,
                     },
-                    awaiting_ack: true,
-                };
+                });
 
                 Ok(vec![DeviceSend::ToUser(DeviceToUserMessage::CheckKeyGen {
                     session_hash,
                 })])
             }
             (
-                SignerState::FrostKey {
-                    key,
-                    awaiting_ack: false,
-                },
+                None,
                 CoordinatorToDeviceMessage::RequestSign(SignRequest {
                     nonces,
                     sign_task,
                     key_id,
                 }),
             ) => {
-                let (my_nonces, my_nonce_index, _) = match nonces.get(&key.share_index) {
-                    Some(nonce) => nonce,
-                    None => return Ok(Vec::new()),
-                };
+                let key = self.keys.get(&key_id).ok_or_else(|| {
+                    Error::signer_invalid_message(
+                        &message,
+                        // we could instead send back a message saying we don't have this key but I
+                        // think this will never happen in practice unless we have a way for one
+                        // coordinator to delete a key from a device without the other coordinator
```

Comment:
I could see this being a possibility in the future, but yes this is good for now!